### ISSUE TITLE
DropInService redirect not working if WebView moved to background

### ISSUE THREAD

**Issue by TheFreshMango** [Adyen/adyen-android#501]
**Description**
In our app, we use the DropInService to process payments. We also use payment methods such as Klarna/Sofort, which open Custom Tabs. The redirect back to the app/Adyen only works if Custom Tabs is not pushed into the background by another app (e.g. banking / to authorize the payment). It is already sufficient to briefly go to the home screen and then return to the open Custom Tabs.

When the payment is authorized via banking app on another smartphone, the redirect works as expected. In this case, Custom Tabs is not moved to the background.

**To Reproduce**
1. Start payment and select a payment method that opens in Custom Tabs (e.g. Klarna/Sofort).
2. Fill in the payment form
3. Before the payment is completed in Custom Tabs, move Custom Tabs to the background (e.g. authorize payment in the banking app or switch to any other app).
4. Complete payment in Custom Tabs

**Expected behavior**
Endless loading indicator inside Custom Tabs, as there is no longer a redirect to the app / Adyen DropInService. Closing the Custom Tabs cancels payment.

**Smartphone:**
 - Device: Samsung SM-N950F (Note 8, European Model)
 - OS: Android 9
 - Adyen: DropInService 4.2.0

** Edit **
The issue might be similar to this one: https://github.com/Adyen/adyen-android/issues/447

**Comment by jreij** [comment]
Hi, thanks for reaching out. Do you have any logs that you can share? To enable logging use `com.adyen.checkout.core.log.Logger.setLogcatLevel(Log.DEBUG)`.

**Comment by TheFreshMango** [comment]
Here are the logs. I recorded these when I performed the payment process on my device and broke them down by step in chronological order for better understanding.

**Step 1: Start Payment**

```
2021-09-28 14:33:56.017 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropIn: startPayment from Activity
2021-09-28 14:33:56.447 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInActivityKt: attachBaseContext
2021-09-28 14:33:56.494 7605-7605/de.myappcafe.myappcafe.myAppCafe_android I/ResourcesCompat: Failed to retrieve rebase() method
    java.lang.NoSuchMethodException: android.content.res.Resources$Theme.rebase []
        at java.lang.Class.getMethod(Class.java:2069)
        at java.lang.Class.getDeclaredMethod(Class.java:2047)
        at androidx.core.content.res.ResourcesCompat$ThemeCompat$ImplApi23.rebase(ResourcesCompat.java:544)
        at androidx.core.content.res.ResourcesCompat$ThemeCompat.rebase(ResourcesCompat.java:520)
        at androidx.appcompat.app.AppCompatDelegateImpl.attachBaseContext2(AppCompatDelegateImpl.java:468)
        at androidx.appcompat.app.AppCompatActivity.attachBaseContext(AppCompatActivity.java:139)
        at com.adyen.checkout.dropin.ui.DropInActivity.attachBaseContext(DropInActivity.kt:135)
        at android.app.Activity.attach(Activity.java:7242)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3057)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3259)
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:78)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1950)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7073)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:494)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:965)
2021-09-28 14:33:56.519 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInActivityKt: onCreate - null
2021-09-28 14:33:56.602 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInActivityKt: showPaymentMethodsDialog
2021-09-28 14:33:56.617 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInActivityKt: handleIntent: action - null
2021-09-28 14:33:56.641 7605-7605/de.myappcafe.myappcafe.myAppCafe_android E/CO.DropInActivityKt: Unable to find action
2021-09-28 14:33:56.641 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInActivityKt: sendAnalyticsEvent
2021-09-28 14:33:56.850 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4f57510[DropInActivity]: setView = DecorView@60bce0c[DropInActivity] TM=true MM=false
2021-09-28 14:33:56.873 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInServiceKt: bindService - DropInActivity
2021-09-28 14:33:56.879 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4f57510[DropInActivity]: setWindowStopped(false) old=false
2021-09-28 14:33:56.895 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4cb2cd4[DropInActivity]: setView = DecorView@679e244[DropInActivity] TM=true MM=false
2021-09-28 14:33:56.931 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4f57510[DropInActivity]: dispatchAttachedToWindow
2021-09-28 14:33:57.290 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4f57510[DropInActivity]: Relayout returned: old=[0,0][1080,2220] new=[0,0][1080,2220] result=0x7 surface={valid=true 511810121728} changed=true
2021-09-28 14:33:57.369 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4cb2cd4[DropInActivity]: dispatchAttachedToWindow
2021-09-28 14:33:57.404 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4cb2cd4[DropInActivity]: Relayout returned: old=[0,0][1080,2220] new=[0,0][1080,2220] result=0x7 surface={valid=true 511526522880} changed=true
2021-09-28 14:33:57.430 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4f57510[DropInActivity]: MSG_WINDOW_FOCUS_CHANGED 1 1
2021-09-28 14:33:57.431 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/InputMethodManager: prepareNavigationBarInfo() DecorView@60bce0c[DropInActivity]
2021-09-28 14:33:57.441 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/InputMethodManager: prepareNavigationBarInfo() DecorView@60bce0c[DropInActivity]
2021-09-28 14:33:57.453 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInServiceKt: onCreate
2021-09-28 14:33:57.455 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInServiceKt: onBind
2021-09-28 14:33:57.526 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4f57510[DropInActivity]: MSG_RESIZED: frame=Rect(0, 0 - 1080, 2220) ci=Rect(0, 63 - 0, 126) vi=Rect(0, 63 - 0, 126) or=1
2021-09-28 14:33:57.530 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4cb2cd4[DropInActivity]: MSG_RESIZED: frame=Rect(0, 0 - 1080, 2220) ci=Rect(0, 63 - 0, 126) vi=Rect(0, 63 - 0, 126) or=1
2021-09-28 14:33:57.537 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInActivityKt: onServiceConnected
```

The payment DropIn is now displayed. It continues with the selection of the payment method.

** Step 2: Selection of Sofort. as payment method (will be opened in Custom Tabs). **

```
2021-09-28 14:34:18.590 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4f57510[DropInActivity]: ViewPostIme pointer 0
2021-09-28 14:34:18.631 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4f57510[DropInActivity]: ViewPostIme pointer 1
2021-09-28 14:34:18.684 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInActivityKt: requestPaymentsCall
2021-09-28 14:34:18.686 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInServiceKt: requestPaymentsCall
2021-09-28 14:34:18.828 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@2dc23bc[DropInActivity]: setView = DecorView@87684b7[DropInActivity] TM=true MM=false
2021-09-28 14:34:18.852 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@2dc23bc[DropInActivity]: dispatchAttachedToWindow
2021-09-28 14:34:18.889 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@2dc23bc[DropInActivity]: Relayout returned: old=[0,63][1080,2094] new=[477,1015][603,1141] result=0x7 surface={valid=true 511448051712} changed=true
2021-09-28 14:34:18.904 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@2dc23bc[DropInActivity]: MSG_WINDOW_FOCUS_CHANGED 1 1
2021-09-28 14:34:18.905 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/InputMethodManager: prepareNavigationBarInfo() DecorView@87684b7[DropInActivity]
2021-09-28 14:34:18.908 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/InputMethodManager: prepareNavigationBarInfo() DecorView@87684b7[DropInActivity]
2021-09-28 14:34:18.969 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@2dc23bc[DropInActivity]: MSG_RESIZED: frame=Rect(477, 1015 - 603, 1141) ci=Rect(0, 0 - 0, 0) vi=Rect(0, 0 - 0, 0) or=1
2021-09-28 14:34:18.981 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4f57510[DropInActivity]: MSG_WINDOW_FOCUS_CHANGED 0 1
2021-09-28 14:34:18.981 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/InputMethodManager: prepareNavigationBarInfo() DecorView@60bce0c[DropInActivity]
2021-09-28 14:34:20.790 7605-7846/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInServiceKt: dispatching DropInServiceResult
2021-09-28 14:34:20.791 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInActivityKt: handleDropInServiceResult - Action
2021-09-28 14:34:20.969 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@2dc23bc[DropInActivity]: MSG_WINDOW_FOCUS_CHANGED 0 1
2021-09-28 14:34:20.970 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/InputMethodManager: prepareNavigationBarInfo() DecorView@87684b7[DropInActivity]
2021-09-28 14:34:21.727 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@2dc23bc[DropInActivity]: Relayout returned: old=[477,1015][603,1141] new=[477,1015][603,1141] result=0x5 surface={valid=false 0} changed=true
2021-09-28 14:34:21.774 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4f57510[DropInActivity]: Relayout returned: old=[0,0][1080,2220] new=[0,0][1080,2220] result=0x5 surface={valid=false 0} changed=true
2021-09-28 14:34:21.791 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@2dc23bc[DropInActivity]: setWindowStopped(true) old=false
2021-09-28 14:34:21.793 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@2dc23bc[DropInActivity]: Surface release. android.view.WindowManagerGlobal.setStoppedState:669 android.app.Activity.performStop:7650 android.app.ActivityThread.callActivityOnStop:4402 android.app.ActivityThread.performStopActivityInner:4380 android.app.ActivityThread.handleStopActivity:4455 android.app.servertransaction.StopActivityItem.execute:41 android.app.servertransaction.TransactionExecutor.executeLifecycleState:145 android.app.servertransaction.TransactionExecutor.execute:70 
2021-09-28 14:34:21.794 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4cb2cd4[DropInActivity]: setWindowStopped(true) old=false
2021-09-28 14:34:21.796 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4cb2cd4[DropInActivity]: Surface release. android.view.WindowManagerGlobal.setStoppedState:669 android.app.Activity.performStop:7650 android.app.ActivityThread.callActivityOnStop:4402 android.app.ActivityThread.performStopActivityInner:4380 android.app.ActivityThread.handleStopActivity:4455 android.app.servertransaction.StopActivityItem.execute:41 android.app.servertransaction.TransactionExecutor.executeLifecycleState:145 android.app.servertransaction.TransactionExecutor.execute:70 
2021-09-28 14:34:21.797 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4f57510[DropInActivity]: setWindowStopped(true) old=false
2021-09-28 14:34:21.799 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4f57510[DropInActivity]: Surface release. android.view.WindowManagerGlobal.setStoppedState:669 android.app.Activity.performStop:7650 android.app.ActivityThread.callActivityOnStop:4402 android.app.ActivityThread.performStopActivityInner:4380 android.app.ActivityThread.handleStopActivity:4455 android.app.servertransaction.StopActivityItem.execute:41 android.app.servertransaction.TransactionExecutor.executeLifecycleState:145 android.app.servertransaction.TransactionExecutor.execute:70 
2021-09-28 14:34:21.819 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInServiceKt: unbindService - DropInActivity
2021-09-28 14:34:21.854 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInActivityKt: onSaveInstanceState
2021-09-28 14:34:21.885 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4cb2cd4[DropInActivity]: Relayout returned: old=[0,0][1080,2220] new=[0,0][1080,2220] result=0x5 surface={valid=false 0} changed=false
2021-09-28 14:34:21.916 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInServiceKt: onUnbind
2021-09-28 14:34:21.917 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInServiceKt: onDestroy
```

I have completed the payment forms and requested payment from my bank. Next I will exit Custom Tabs to open my banking app.

** Step 3: Switch to the banking app **

There are no logs for this. After the payment is authorized, I switch back to Custom Tabs. Payment was successful but no redirect happens, instead only the endless loading screen is displayed.

** Step 4: Workaround I discovered ** 

When I open the options menu in Custom Tabs, I can view the current page in Samsung Browser. The page is opened there without redirect (loading screen). Pressing back closes Samsung Browser and reopens Custom Tabs. Now the redirect is done, but not always.
After redirect there are new logs and the payment was successful:

```
2021-09-28 14:38:01.763 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4f57510[DropInActivity]: Relayout returned: old=[0,0][1080,2220] new=[0,0][1080,2220] result=0x1 surface={valid=false 0} changed=false
2021-09-28 14:38:01.794 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4cb2cd4[DropInActivity]: Relayout returned: old=[0,0][1080,2220] new=[0,0][1080,2220] result=0x1 surface={valid=false 0} changed=false
2021-09-28 14:38:01.816 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInActivityKt: onNewIntent
2021-09-28 14:38:01.816 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInActivityKt: handleIntent: action - android.intent.action.VIEW
2021-09-28 14:38:01.830 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@2dc23bc[DropInActivity]: setWindowStopped(false) old=true
2021-09-28 14:38:01.830 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4cb2cd4[DropInActivity]: setWindowStopped(false) old=true
2021-09-28 14:38:01.831 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4f57510[DropInActivity]: setWindowStopped(false) old=true
2021-09-28 14:38:01.842 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInServiceKt: bindService - DropInActivity
2021-09-28 14:38:01.849 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInActivityKt: requestDetailsCall
2021-09-28 14:38:01.849 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInServiceKt: requestDetailsCall
2021-09-28 14:38:01.856 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@2dc23bc[DropInActivity]: setWindowStopped(false) old=false
2021-09-28 14:38:01.856 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4cb2cd4[DropInActivity]: setWindowStopped(false) old=false
2021-09-28 14:38:01.857 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4f57510[DropInActivity]: setWindowStopped(false) old=false
2021-09-28 14:38:01.914 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@2dc23bc[DropInActivity]: Relayout returned: old=[477,1015][603,1141] new=[477,1015][603,1141] result=0x7 surface={valid=true 511810121728} changed=true
2021-09-28 14:38:01.980 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4cb2cd4[DropInActivity]: Relayout returned: old=[0,0][1080,2220] new=[0,0][1080,2220] result=0x7 surface={valid=true 511996350464} changed=true
2021-09-28 14:38:02.048 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4f57510[DropInActivity]: Relayout returned: old=[0,0][1080,2220] new=[0,0][1080,2220] result=0x7 surface={valid=true 511448403968} changed=true
2021-09-28 14:38:02.098 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInServiceKt: onCreate
2021-09-28 14:38:02.099 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInServiceKt: onBind
2021-09-28 14:38:02.222 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@2dc23bc[DropInActivity]: MSG_WINDOW_FOCUS_CHANGED 1 1
2021-09-28 14:38:02.295 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/InputMethodManager: prepareNavigationBarInfo() DecorView@87684b7[DropInActivity]
2021-09-28 14:38:02.298 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/InputMethodManager: prepareNavigationBarInfo() DecorView@87684b7[DropInActivity]
2021-09-28 14:38:02.308 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInActivityKt: onServiceConnected
2021-09-28 14:38:08.969 7605-7846/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInServiceKt: dispatching DropInServiceResult
2021-09-28 14:38:08.970 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInActivityKt: handleDropInServiceResult - Finished
2021-09-28 14:38:08.972 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInActivityKt: terminateSuccessfully
2021-09-28 14:38:08.972 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInActivityKt: terminate
2021-09-28 14:38:09.098 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@2dc23bc[DropInActivity]: MSG_WINDOW_FOCUS_CHANGED 0 1
2021-09-28 14:38:09.098 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/InputMethodManager: prepareNavigationBarInfo() DecorView@87684b7[DropInActivity]
2021-09-28 14:38:10.270 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@2dc23bc[DropInActivity]: Relayout returned: old=[477,1015][603,1141] new=[477,1015][603,1141] result=0x5 surface={valid=false 0} changed=true
2021-09-28 14:38:10.322 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4cb2cd4[DropInActivity]: Relayout returned: old=[0,0][1080,2220] new=[0,0][1080,2220] result=0x5 surface={valid=false 0} changed=true
2021-09-28 14:38:10.375 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4f57510[DropInActivity]: Relayout returned: old=[0,0][1080,2220] new=[0,0][1080,2220] result=0x5 surface={valid=false 0} changed=true
2021-09-28 14:38:10.708 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@2dc23bc[DropInActivity]: setWindowStopped(true) old=false
2021-09-28 14:38:10.712 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@2dc23bc[DropInActivity]: Surface release. android.view.WindowManagerGlobal.setStoppedState:669 android.app.Activity.performStop:7650 android.app.ActivityThread.callActivityOnStop:4402 android.app.ActivityThread.performStopActivityInner:4380 android.app.ActivityThread.handleStopActivity:4455 android.app.servertransaction.TransactionExecutor.performLifecycleSequence:192 android.app.servertransaction.TransactionExecutor.cycleToPath:165 android.app.servertransaction.TransactionExecutor.executeLifecycleState:142 
2021-09-28 14:38:10.712 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4cb2cd4[DropInActivity]: setWindowStopped(true) old=false
2021-09-28 14:38:10.717 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4cb2cd4[DropInActivity]: Surface release. android.view.WindowManagerGlobal.setStoppedState:669 android.app.Activity.performStop:7650 android.app.ActivityThread.callActivityOnStop:4402 android.app.ActivityThread.performStopActivityInner:4380 android.app.ActivityThread.handleStopActivity:4455 android.app.servertransaction.TransactionExecutor.performLifecycleSequence:192 android.app.servertransaction.TransactionExecutor.cycleToPath:165 android.app.servertransaction.TransactionExecutor.executeLifecycleState:142 
2021-09-28 14:38:10.718 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4f57510[DropInActivity]: setWindowStopped(true) old=false
2021-09-28 14:38:10.720 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4f57510[DropInActivity]: Surface release. android.view.WindowManagerGlobal.setStoppedState:669 android.app.Activity.performStop:7650 android.app.ActivityThread.callActivityOnStop:4402 android.app.ActivityThread.performStopActivityInner:4380 android.app.ActivityThread.handleStopActivity:4455 android.app.servertransaction.TransactionExecutor.performLifecycleSequence:192 android.app.servertransaction.TransactionExecutor.cycleToPath:165 android.app.servertransaction.TransactionExecutor.executeLifecycleState:142 
2021-09-28 14:38:10.741 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInServiceKt: unbindService - DropInActivity
2021-09-28 14:38:10.803 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4f57510[DropInActivity]: dispatchDetachedFromWindow
2021-09-28 14:38:10.804 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4f57510[DropInActivity]: Surface release. android.view.ViewRootImpl.doDie:7964 android.view.ViewRootImpl.die:7932 android.view.WindowManagerGlobal.removeViewLocked:497 android.view.WindowManagerGlobal.removeView:435 android.view.WindowManagerImpl.removeViewImmediate:124 android.app.Dialog.dismissDialog:518 android.app.Dialog.dismiss:501 androidx.appcompat.app.AppCompatDialog.dismiss:140 
2021-09-28 14:38:10.839 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@2dc23bc[DropInActivity]: dispatchDetachedFromWindow
2021-09-28 14:38:10.840 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@2dc23bc[DropInActivity]: Surface release. android.view.ViewRootImpl.doDie:7964 android.view.ViewRootImpl.die:7932 android.view.WindowManagerGlobal.removeViewLocked:497 android.view.WindowManagerGlobal.removeView:435 android.view.WindowManagerImpl.removeViewImmediate:124 android.app.Dialog.dismissDialog:518 android.app.Dialog.dismiss:501 androidx.fragment.app.DialogFragment.onDestroyView:747 
2021-09-28 14:38:10.871 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4cb2cd4[DropInActivity]: dispatchDetachedFromWindow
2021-09-28 14:38:10.872 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/ViewRootImpl@4cb2cd4[DropInActivity]: Surface release. android.view.ViewRootImpl.doDie:7964 android.view.ViewRootImpl.die:7932 android.view.WindowManagerGlobal.removeViewLocked:497 android.view.WindowManagerGlobal.removeView:435 android.view.WindowManagerImpl.removeViewImmediate:124 android.app.ActivityThread.handleDestroyActivity:4776 android.app.servertransaction.DestroyActivityItem.execute:39 android.app.servertransaction.TransactionExecutor.executeLifecycleState:145 
2021-09-28 14:38:10.885 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInServiceKt: onUnbind
2021-09-28 14:38:10.886 7605-7605/de.myappcafe.myappcafe.myAppCafe_android D/CO.DropInServiceKt: onDestroy
```

**Comment by jreij** [comment]
Hmm strange indeed. 

> After the payment is authorized, I switch back to Custom Tabs

Do you switch back manually? The banking app should redirect you automatically, if that is not happening then maybe there is a problem with your redirect URL. Can you check the logs you sent me and try to find something about that URL? Specifically `CO.RedirectDelegateKt: makeRedirect - YOUR_URL` and `CO.RedirectUtil: parseRedirectResult - ANOTHER_URL`.

**Comment by jreij** [comment]
Ah I see. And what about `CO.RedirectUtil: parseRedirectResult - URL`? Can you see that after your workaround (Step 4)?

**Comment by TheFreshMango** [comment]
Sorry my bad, here is the part after redirecting:
```
2021-09-28 17:01:38.849 30073-30073/de.myappcafe.myappcafe.myAppCafe_android D/CO.ActionHandler: handleAction - loaded component type: RedirectComponent
2021-09-28 17:01:38.849 30073-30073/de.myappcafe.myappcafe.myAppCafe_android D/CO.RedirectUtil: parseRedirectResult - adyencheckout://de.myappcafe.myappcafe.myAppCafe_android?redirectResult=X6XtfGC3%21eyJyZWRpcmVjdFBheW1lbnREYXRhIjp7ImFjcXVpcmVyQWNjb3VudElkIjoxNTQ0MDY4LCJhbW91bnQiOnsiY3VycmVuY3kiOiJFVVIiLCJ2YWx1ZSI6MzAwfSwiYXBpVmVyc2lvbiI6NjcsImFwcGxpY2F0aW9uSW5mbyI6eyJhZHllbkxpYnJhcnkiOnsibmFtZSI6ImFkeWVuLW5vZGUtYXBpLWxpYnJhcnkiLCJ2ZXJzaW9uIjoiOC4yLjAifX0sImJyYW5kQ29kZSI6ImRpcmVjdEViYW5raW5nIiwiY291bnRyeUNvZGUiOiJERSIsIm1lcmNoYW50QWNjb3VudCI6Ik15QXBwQ2FmZUdtYkhFQ09NIiwibWVyY2hhbnRSZWZlcmVuY2UiOiJ6QmdsRG1DZSIsIm9yaWdpbmFsUHNwUmVmZXJlbmNlIjoiMDExNjMyODQxMjExMDcxNyIsInBhbFJlc3VsdFBzcFJlZmVyZW5jZSI6IjQ4MTYzMjg0MTIxMTg5NjIiLCJyZXF1ZXN0QWNxdWlyZXJSZWZlcmVuY2UiOiIxMjAxMDAtNzM4NzI1LTYxNTMyREZCLUQyRTEiLCJyZXF1ZXN0QWRkaXRpb25hbERhdGEiOnsiYXBwbGljYXRpb25JbmZvLmxpYnJhcnlOYW1lIjoiYWR5ZW4tbm9kZS1hcGktbGlicmFyeSIsImFwcGxpY2F0aW9uSW5mby5saWJyYXJ5VmVyc2lvbiI6IjguMi4wIiwibWVyY2hhbnRJbnRlZ3JhdGlvbi50eXBlIjoiQ0hFQ0tPVVRfR0VORVJJQyIsIm1lcmNoYW50SW50ZWdyYXRpb24udmVyc2lvbiI6IjY3In0sInJldHVyblVybCI6Imh0dHBzOlwvXC9jaGVja291dHNob3BwZXItbGl2ZS5hZHllbi5jb21cL2NoZWNrb3V0c2hvcHBlclwvY2hlY2tvdXRQYXltZW50UmV0dXJuP2dwaWQ9R1A5RjU4NzhFNDQzMzlGNDBBIiwic2VsZWN0ZWRCcmFuZCI6ImRpcmVjdEViYW5raW5nIiwic2hvcHBlckxvY2FsZSI6IkRFIiwic2tpbkNvZGUiOiJwdWIudjIuNDExNjMyMjg2NTEyMzcxNy5xNGlfSjVWN0M0aXlUVjFNQTRWTDd2UmVUdl8tbWRQQm1CTWNRVlM4em1VIn0sInJldHVyblVybFBhcmFtcyI6eyJEVU1NRVlfS0VZIjoiRFVNTUVZX1ZBTFVFIn19UfxDRElDuyCi6hBjj%2F0vzhSYQBT0VWB3chg%2BbVLyXNM%3D
```

**Comment by jreij** [comment]
So after investigating more, it seems like we have seen similar issues before and it's probably on Klarna's side. What we think is happening is that Klarna's webpage is trying to make the redirect while the user is still inside their banking app, but Android blocks these redirects because the app is in the background. Now when the user goes back to your app, the Klarna page should retry to redirect or present a button for the user to manually proceed with the redirect. But since it doesn't, the redirect never actually happens.

If you'd like to look into this furthermore, you can [debug your phone's browser on your machine](https://developer.chrome.com/docs/devtools/remote-debugging/) and you should see an error log from the Klarna web page shortly after the payment is confirmed in the banking app.

Ultimately if that is the issue indeed, it can only be solved by Klarna. Maybe you could contact them and ask if they have the option to add a button on the final loading screen so that the user retries.

**Comment by nilskasseckert** [comment]
Hi @jreij 
I am a colleague of @TheFreshMango and will now take over.

I personally can not imagine a Klarna problem. We have also implemented the iOS SDK. There is this button. (see picture). The question is: Why does it look different on Android than on iOS?

Furthermore, it will not bring anything if we deal with Klarna. You provide the SDK and have a corresponding size. Klarna would only put us off. Furthermore, you also offer Klarna and your SDK is unfortunately almost unusable in Germany, since only Klarna/PayPal is actually used in addition to Credit Card.  

Regards,
Nils

**Comment by jreij** [comment]
Hi again, we're looking into this issue with Klarna. However as the logs we have at this point are not sufficient, we'll need some extra information to better troubleshoot it.

Please reach out to support@adyen.com explaining the problem, and providing the [link to this GitHub issue](https://github.com/Adyen/adyen-android/issues/501) as well as a sample [PSP reference](https://docs.adyen.com/get-started-with-adyen/payment-glossary#psp-reference) of one of the payments that did not trigger a redirect back to your app. Then we'll be able to assist you furthermore!

**Comment by nilskasseckert** [comment]
Hi,
we will open a ticket but we have also investigated it further. And it can actually only be a problem in the library.

Following use case, same problem:

Given: Mobile 1 (M1), Mobile 2 (M2).

Sequence:
1. M1 starts SOFORT transfer
2. M1 logs in at his bank
3. M2 confirms the login request
3. M1 pushes the app into the background 
4. M1 gets the app out of the background
5. M1 completes the payment
6. M2 confirms the payment, H1 app is meanwhile in the foreground 
--> endless loading comes

**Comment by jreij** [comment]
Hi, thanks for opening a ticket and for the extra investigation, very helpful indeed!

Just to get a bit more insight, does anything particular happen between step 4 and step 5, on either M1 or M2? Or is putting the app in the background at any stage during the payment process enough to cause the issue?

**Comment by nilskasseckert** [comment]
Hi,
Putting the App in the background at any stage is enough.

Regards,
Nils

**Comment by jreij** [comment]
Hi again, we were able to reproduce the issue and found out that it was indeed caused by a redirect being blocked while the app is in the background.

Klarna confirmed the issue from their side and informed us that they released a fix by adding a button (underlined text) that the user can click in case the redirect fails in the background. We tested it on our side also and were able to confirm the fix.

Let us know if you encounter this again.
